### ISSUE TITLE
Provide a hook when runtimeConfig variable is resolved

### ISSUE THREAD

**Issue by noook** [nuxt/nuxt#20561]
### Describe the feature

Given the documentation [here](https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables)

> Your desired variables must be defined in your nuxt.config. This ensures that arbitrary environment variables are not exposed to your application code.

> Only a specially-named environment variable can override a runtime config property. That is, an uppercase environment variable starting with NUXT_ which uses _ to separate keys and case changes.

[Example](https://nuxt.com/docs/guide/going-further/runtime-config#example)

#### `.env`

```env
NUXT_API_SECRET=api_secret_token
NUXT_PUBLIC_API_BASE=https://nuxtjs.org
```

#### `nuxt.config.ts`
```ts
export default defineNuxtConfig({
  runtimeConfig: {
    apiSecret: '', // can be overridden by NUXT_API_SECRET environment variable
    public: {
      apiBase: '', // can be overridden by NUXT_PUBLIC_API_BASE environment variable
    }
  },
})
```

I am not sure it is possible to sanitize the environment variables when they are resolved so that we don't need to sanitize them everywhere. I tried using hooks so I could read the value, sanitize it, then reassign it, but I'm not sure it works that way. I have `NUXT_PUBLIC_PLATFORM_URL` defined and set.

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  hooks: {
    ready(nuxt) {
      console.log({ url: nuxt.options.runtimeConfig.public.platformUrl }) // resolves to empty string
      // Sanitize here
      nuxt.options.runtimeConfig.public.platformUrl = trimTrailingSlash(nuxt.options.runtimeConfig.public.platformUrl)
    }
  }
})
```

This hook says that the nuxt instance is ready to work, but the env variables seem not to be read at this moment. Would it be useful to provide a hook for when the env variables resolve, so we could sanitize them ?

```ts
// nuxt.config.ts
export default defineNuxtConfig({
  hooks: {
    'env:resolve'(env: Record<string, string>) {
      env.NUXT_PUBLIC_PLATFORM_URL = trimTrailingSlash(env.NUXT_PUBLIC_PLATFORM_URL);

      return env;
    }
  }
})
```

### Additional information

- [X] Would you be willing to help implement this feature?
- [ ] Could this feature be implemented as a module?

### Final checks

- [X] Read the [contribution guide](https://nuxt.com/docs/community/contribution).
- [X] Check existing [discussions](https://github.com/nuxt/nuxt/discussions) and [issues](https://github.com/nuxt/nuxt/issues).

**Comment by danielroe** [comment]
We can certainly implement this as a build/dev time hook both within Nuxt and Nitro (@pi0 and I will align to ensure the implementation is compatible).

Does that meet your use case or do you think there's also a need for a runtime hook?

**Comment by noook** [comment]
Given they are environment variables, I don't think they can change at runtime, so unless someone has this requirement I think build/dev time should be enough

**Comment by wolodek** [comment]
+1, I've been trying to find solution for this too. In Nuxt 2 I handled it via runtime module but it's impossible in Nuxt 3 :/

**Comment by Hebilicious** [comment]
> It looks like someone else is having exactly the same requirement than us: [stackoverflow.com/questions/76488291/how-to-fetch-data-as-part-of-server-start-up-in-nuxt-3](https://stackoverflow.com/questions/76488291/how-to-fetch-data-as-part-of-server-start-up-in-nuxt-3)
> 
> So it's definitely a runtime hook, maybe `before:ready`

If you can shift that logic at build time, it would be easy to achieve with a module.

```TS
const configKey = "yourKey"

export default defineNuxtModule({
  meta: {
    name: "config-environment",
  },
  async setup(userOptions, nuxt) {
    const response = await fetch(stuff)
    const options = extractOptions(response.data)
    nuxt.options.runtimeConfig.public[configKey] = options
  }
})
```

If your docker requirements, given that you can just use the `NUXT_` prefix for your environments variables and access them, I assume that you just want them to be sanitized ? 
In this case, maybe we should just sanitize all environments variables by default, so that you end-users don't need additional logic. @danielroe  Is there any reason to not do this by default ?

**Comment by wolodek** [comment]
@Hebilicious I guess that's the whole point to do it on runtime. If it was done in built time then we would have to ship whole repo to each env, doing npm installs (with installing dev depencencies which are often full of vulnerabilities ) and build. 
From CI/CD perspective it's much more optimal to have app builded only once and then ship small package anywhere.

**Comment by Hebilicious** [comment]
> @Hebilicious I guess that's the whole point to do it on runtime. If it was done in built time then we would have to ship whole repo to each env, doing npm installs (with installing dev depencencies which are often full of vulnerabilities ) and build. From CI/CD perspective it's much more optimal to have app builded only once and then ship small package anywhere.

It depends on your setup. For node.js, you could argue that installing deps and building is more optimal that sending large folders of built files that contains node modules anyways. You could also run into issue where your build env and your prod env are incompatible etc.

@noook said initially : "Given they are environment variables, I don't think they can change at runtime, so unless someone has this requirement I think build/dev time should be enough"; hence my answer.

@wolodek are you trying to sanitise environments variables or are you trying to achieve something else?

**Comment by wolodek** [comment]
@Hebilicious Sanitising and validating is what we need right now but similarly to @noook we would like to fetch config in future.

**Comment by noook** [comment]
@Hebilicious It was indeed my first requirement, but we figured we would also need the possibility to fetch values before allowing requests to the server.

I imagined a hook, but the hook could serve many objectives, such as sanitizing environment variables, or fetching data that serves as runtime config

**Comment by Hebilicious** [comment]
@danielroe I think that for build/dev time this is un-necessary as this is easily done as a module.

I suggest  :

- We automatically sanitize environment variables (does this happens only here ? https://github.com/unjs/nitro/blob/main/src/runtime/config.ts)

- We add a new nitro runtime hook that runs within `createNitroApp`, or we expose the nitro config to the nitro plugins, which would allow nitro plugins to modify it, either through a hook or directly within the plugin. (I'm assuming that Nitro plugins only run one time when the server start if you're not using a serverless environment ?)

And then the public variables are passed from Nitro to Nuxt through the renderer.

**Comment by Hebilicious** [comment]
Might be related to https://volta.net/unjs/nitro/issues/339

**Comment by Norbiros** [comment]
Probably a duplicate of https://github.com/nuxt/nuxt/issues/15366. It is better to track it in one place

**Comment by danielroe** [comment]
nuxtathon closed @Norbiros
